File: library/src/main/java/com/cosic/instagallery/gallery/GalleryRecyclerView.java
package com.cosic.instagallery.gallery;

import android.content.Context;
import android.graphics.PointF;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.Nullable;
import androidx.core.view.ViewCompat;
import androidx.recyclerview.widget.LinearSmoothScroller;
import androidx.recyclerview.widget.RecyclerView;
import com.cosic.instagallery.utils.Logger;
import com.google.android.material.appbar.AppBarLayout;

public class GalleryRecyclerView extends RecyclerView {

    @Nullable
    private OnDispatchTouchListener mOnDispatchTouchListener;

    private RecyclerView.SmoothScroller mSmoothScroller;

    private AppBarLayout mChild;

    public GalleryRecyclerView(final Context context) {
        this(context, null);
    }

    public GalleryRecyclerView(final Context context, @Nullable final AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public GalleryRecyclerView(final Context context, @Nullable final AttributeSet attrs, final int defStyle) {
        super(context, attrs, defStyle);
        mSmoothScroller = new GallerySmoothScroller(context);
    }

    @Override
    protected void onLayout(final boolean changed, final int l, final int t, final int r, final int b) {
        super.onLayout(changed, l, t, r, b);
        AppBarLayout child = getNestedAppBar();
        if (child != null) {
            int childTop = child.getTop();
            int childHeight = child.getHeight();
            int childBottom = child.getBottom();
            int top = getTop();
            Logger.d("onLayout, childBottom=%d childTop=%d, childHeight=%d, top=%d",
                    childBottom, childTop, childHeight, top);
            ViewCompat.offsetTopAndBottom(this, childBottom - top);
//            ViewGroup.LayoutParams lp = getLayoutParams();
//            lp.height += child.getHeight();
//            setLayoutParams(lp);
        }
    }

    private AppBarLayout getNestedAppBar() {
        if (mChild == null) {
            ViewGroup parent = (ViewGroup) getParent();
            for (int i = 0; i < parent.getChildCount(); i++) {
                View childAt = parent.getChildAt(i);
                if (childAt instanceof AppBarLayout) {
                    mChild = (AppBarLayout) childAt;
                }
            }
        }
        return mChild;
    }

    public void scrollToPosition(int position) {
        LayoutManager lm = getLayoutManager();
        if (lm == null) return;
        mSmoothScroller.setTargetPosition(position);
        lm.startSmoothScroll(mSmoothScroller);
    }

    @Override
    public boolean onTouchEvent(final MotionEvent e) {
        if (mOnDispatchTouchListener != null) {
            if (mOnDispatchTouchListener.onDispatchTouchEvent(this, e)) {
                return true;
            }
        }

        switch (e.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                Logger.d("GalleryRecyclerView: onTouchEvent ACTION_DOWN");
                break;
            case MotionEvent.ACTION_MOVE:
                break;
            case MotionEvent.ACTION_CANCEL:
            case MotionEvent.ACTION_UP:
                Logger.d("GalleryRecyclerView: onTouchEvent ACTION_UP");
                break;
        }
        return super.onTouchEvent(e);
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent e) {
        switch (e.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                if (mOnDispatchTouchListener != null) {
                    mOnDispatchTouchListener.onDispatchTouchEvent(this, e);
                }
                Logger.d("GalleryRecyclerView: dispatchTouchEvent ACTION_DOWN");
                break;
            case MotionEvent.ACTION_MOVE:
                break;
            case MotionEvent.ACTION_CANCEL:
            case MotionEvent.ACTION_UP:
                Logger.d("GalleryRecyclerView: dispatchTouchEvent ACTION_UP");
                break;
        }
        return super.dispatchTouchEvent(e);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent e) {
        switch (e.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                Logger.d("GalleryRecyclerView: onInterceptTouchEvent ACTION_DOWN");
                break;
            case MotionEvent.ACTION_MOVE:
                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                Logger.d("GalleryRecyclerView: onInterceptTouchEvent ACTION_UP");
                break;
        }
        return super.onInterceptTouchEvent(e);
    }

    public void setOnDispatchTouchListener(@Nullable final OnDispatchTouchListener listener) {
        mOnDispatchTouchListener = listener;
    }

    public interface OnDispatchTouchListener {
        boolean onDispatchTouchEvent(View view, MotionEvent e);
    }

    public static class GallerySmoothScroller extends LinearSmoothScroller {

        private static final float MILLISECONDS_PER_INCH = 50f; //default is 25f (bigger == slower)

        public GallerySmoothScroller(final Context context) {
            super(context);
        }

        @Override protected int getVerticalSnapPreference() {
            return LinearSmoothScroller.SNAP_TO_START;
        }

        @Override
        public PointF computeScrollVectorForPosition(int targetPosition) {
            return super.computeScrollVectorForPosition(targetPosition);
        }

        @Override
        protected float calculateSpeedPerPixel(DisplayMetrics displayMetrics) {
            return MILLISECONDS_PER_INCH / displayMetrics.densityDpi;
        }
    }
}

File: library/src/main/java/com/cosic/instagallery/GalleryRecyclerView.kt
package com.cosic.instagallery

import android.content.Context
import android.graphics.PointF
import android.util.AttributeSet
import android.util.DisplayMetrics
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.LinearSmoothScroller
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.appbar.AppBarLayout

class GalleryRecyclerView
@JvmOverloads
constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : RecyclerView(context, attrs, defStyle) {

    private var mOnDispatchTouchListener: OnDispatchTouchListener? = null

    private val mSmoothScroller: SmoothScroller

    private var mChild: AppBarLayout? = null

    private val nestedAppBar: AppBarLayout?
        get() {
            if (mChild == null) {
                val parent = parent as ViewGroup
                for (i in 0 until parent.childCount) {
                    val childAt = parent.getChildAt(i)
                    if (childAt is AppBarLayout) {
                        mChild = childAt
                    }
                }
            }
            return mChild
        }

    init {
        mSmoothScroller = GallerySmoothScroller(context)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
        val child = nestedAppBar
        if (child != null) {
            val childTop = child.top
            val childHeight = child.height
            val childBottom = child.bottom
            val top = top
            Logger.d("onLayout, childBottom=%d childTop=%d, childHeight=%d, top=%d",
                childBottom, childTop, childHeight, top)
            ViewCompat.offsetTopAndBottom(this, childBottom - top)
            //            ViewGroup.LayoutParams lp = getLayoutParams();
            //            lp.height += child.getHeight();
            //            setLayoutParams(lp);
        }
    }

    override fun scrollToPosition(position: Int) {
        val lm = layoutManager ?: return
        mSmoothScroller.targetPosition = position
        lm.startSmoothScroll(mSmoothScroller)
    }

    override fun onTouchEvent(e: MotionEvent): Boolean {
        if (mOnDispatchTouchListener != null) {
            if (mOnDispatchTouchListener!!.onDispatchTouchEvent(this, e)) {
                return true
            }
        }

        when (e.actionMasked) {
            MotionEvent.ACTION_DOWN -> Logger.d("GalleryRecyclerView: onTouchEvent ACTION_DOWN")
            MotionEvent.ACTION_MOVE -> {
            }
            MotionEvent.ACTION_CANCEL, MotionEvent.ACTION_UP -> Logger.d("GalleryRecyclerView: onTouchEvent ACTION_UP")
        }
        return super.onTouchEvent(e)
    }

    override fun dispatchTouchEvent(e: MotionEvent): Boolean {
        when (e.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                if (mOnDispatchTouchListener != null) {
                    mOnDispatchTouchListener!!.onDispatchTouchEvent(this, e)
                }
                Logger.d("GalleryRecyclerView: dispatchTouchEvent ACTION_DOWN")
            }
            MotionEvent.ACTION_MOVE -> {
            }
            MotionEvent.ACTION_CANCEL, MotionEvent.ACTION_UP -> Logger.d("GalleryRecyclerView: dispatchTouchEvent ACTION_UP")
        }
        return super.dispatchTouchEvent(e)
    }

    override fun onInterceptTouchEvent(e: MotionEvent): Boolean {
        when (e.actionMasked) {
            MotionEvent.ACTION_DOWN -> Logger.d("GalleryRecyclerView: onInterceptTouchEvent ACTION_DOWN")
            MotionEvent.ACTION_MOVE -> {
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> Logger.d("GalleryRecyclerView: onInterceptTouchEvent ACTION_UP")
        }
        return super.onInterceptTouchEvent(e)
    }

    fun setOnDispatchTouchListener(listener: OnDispatchTouchListener?) {
        mOnDispatchTouchListener = listener
    }

    interface OnDispatchTouchListener {
        fun onDispatchTouchEvent(view: View, e: MotionEvent): Boolean
    }

    class GallerySmoothScroller(context: Context) : LinearSmoothScroller(context) {

        override fun getVerticalSnapPreference(): Int {
            return SNAP_TO_START
        }

        override fun computeScrollVectorForPosition(targetPosition: Int): PointF? {
            return super.computeScrollVectorForPosition(targetPosition)
        }

        override fun calculateSpeedPerPixel(displayMetrics: DisplayMetrics): Float {
            return MILLISECONDS_PER_INCH / displayMetrics.densityDpi
        }

        companion object {

            private val MILLISECONDS_PER_INCH = 50f //default is 25f (bigger == slower)
        }
    }
}
